cohttp-lwt/src/client.ml

```diff
@@ -13,33 +13,18 @@ module Make
   type ctx = Net.ctx [@@deriving sexp_of]
   let default_ctx = Net.default_ctx
 
-  let read_response ~closefn ic _oc meth =
-    Response.read ic >>= begin function
-    | `Invalid reason ->
-      Lwt.fail (Failure ("Failed to read response: " ^ reason))
-    | `Eof -> Lwt.fail (Failure "Client connection was closed")
-    | `Ok res -> begin
-        let has_body = match meth with
-          | `HEAD -> `No
-          | _ -> Response.has_body res
-        in
-        match has_body with
-        | `Yes | `Unknown ->
-          let reader = Response.make_body_reader res ic in
-          let stream = Body.create_stream Response.read_body_chunk reader in
-          let closefn = closefn in
-          Lwt.on_success (Lwt_stream.closed stream) closefn;
-          let gcfn _st = closefn () in
-          Gc.finalise gcfn stream;
-          let body = Body.of_stream stream in
-          Lwt.return (res, body)
-        | `No -> closefn (); Lwt.return (res, `Empty)
-      end
-    end
-    |> fun t ->
-    Lwt.on_cancel t closefn;
-    Lwt.on_failure t (fun _exn -> closefn ());
-    t
+  let read_body ~closefn ic res =
+    match Response.has_body res with
+    | `Yes | `Unknown ->
+      let reader = Response.make_body_reader res ic in
+      let stream = Body.create_stream Response.read_body_chunk reader in
+      let body = Body.of_stream stream in
+      Lwt.on_success (Lwt_stream.closed stream) closefn;
+      Gc.finalise
+        (fun body -> Lwt.async (fun () -> Body.drain_body body))
```

Comment:
The finaliser may run in any (system) thread. `Body.drain_body` calls into Lwt and therefore must be run in the Lwt main thread. How can this be accomplished? I suspect `Lwt_preemptive.run_in_main` is no option because not available in mirage !? Any suggestions how to deal with this?

Reply:
To me it seems that 589464a is the most reasonable thing to do, given the constraints. But I'll wait for a more mirage-knowledgeable reviewer before taking premature decisions

Reply:
I can confirm that `Lwt_preemptive.run_in_main` is only available in Unix and would be a problem for mirage

Reply:
Lwt_preemptive is not available in MirageOS, sorry about that. I'm not entirely sure what is the best to do here.

Reply:
No problem, thanks for looking into it